osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/messaging/DlmsConnectionMessageProcessor.java

```diff
@@ -147,8 +147,12 @@ protected void doConnectionPostProcessing(
     }
 
     if (device.needsInvocationCounter()) {
-      this.updateInvocationCounterForDevice(device, conn);
-
+      final boolean invocationCounterLowered =
+          this.systemEventService.receivedInvocationCounterIsLowerThanCurrentValue(
+              device, messageMetadata, conn);
```

Comment:
This looks like the wrong place to perform this check.

One reason would be that it does not belong as a check with all connections an invocation counter is involved, but only in those connections where GXF needs to update the invocation counter before getting a working connection (that is, where GXF can be manipulated to reuse invocation counter values).

Another issue is related to the implementation of the check, where device communication is used to retrieve the invocation counter. Besides my objections to this practice, as expressed in the `SystemEventService` review comments, there is an issue that the throttling constraints are ignored. You should not communicate with a device after releasing the throttling permit, as is done a couple of lines above this change.

Reply:
In the code there are flows from
`org.opensmartgridplatform.adapter.protocol.dlms.domain.factories.DlmsConnectionHelper.createAndHandleConnectionForDevice(DlmsDevice, DlmsMessageListener, ConnectionProperties, MessageMetadata, Permit, Consumer<DlmsConnectionManager>)`
where the invocation counter is out of sync, which is then handled by:
`org.opensmartgridplatform.adapter.protocol.dlms.domain.factories.InvocationCounterManager.initializeWithInvocationCounterStoredOnDeviceTask(DlmsDevice, DlmsConnectionManager)`.
Here you'll  find:
```java
      final Long previousKnownInvocationCounter = device.getInvocationCounter();
      final Long invocationCounterFromDevice = this.getInvocationCounter(connectionManager);
```
What I would expected to see was that in this method a check would be performed if `invocationCounterFromDevice` is a lower value than `previousKnownInvocationCounter`. If so, the value should not be set on the device, in order to prevent it from being used (with the risk of reusing an invocation counter).
Probably some kind of exception should be thrown here to terminate the attempt to create the device connection.

Then, at some appropriate place in the flow the system event could be sent, and the request being handled could get an error response.
Maybe this should not even be a system event, as these are usually used to signal something separate from the web service request-response flow. In this case I would think a functional exception code could be more appropriate, where the application could translate this in an event code for other systems.